subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotterTest.groovy

```diff
@@ -243,6 +248,39 @@ class DefaultFileSystemSnapshotterTest extends Specification {
         })
     }
 
+    def "snapshots a singletonFileTree as RegularFileSnapshot"() {
+        given:
+        def file = tmpDir.createFile('testFile')
+
+        when:
+        def tree = new FileTreeAdapter(new DefaultSingletonFileTree(file), TestFiles.patternSetFactory)
+        def snapshots = snapshotter.snapshot(tree)
+
+        then:
+        snapshots.size() == 1
+        getSnapshotInfo(snapshots[0]) == [null, 1]
+    }
+
+    def "snapshots a generated singletonFileTree as RegularFileSnapshot"() {
+        given:
+        def file = tmpDir.createFile('testFile')
+        def tmpDir = tmpDir.createDir('tmpDir')
+
+        when:
+        def tree = new FileTreeAdapter(new GeneratedSingletonFileTree(tmpDir, TestFiles.directoryFileTreeFactory(), file.name, new Action<OutputStream>() {
```

Comment:
We also need tests for filtered versions of this file tree.

Reply:
I added tests for that. But turns out adding filters here creates a new FileTree which then falls back calling AbstractFileTree#visitLeafCollections instead of FileTreeAdapter#visitLeafCollections. I'm not sure what the expected behaviour should be tbh. WDYT?